iox-#743 Validate callbacks in c example (#1188)
eclipse-iceoryx/iceoryx: iceoryx_examples/callbacks_in_c/README.md

For a general introduction into the Listener concept please take a look at
the first part of the

> [Listener C++ example](https://github.com/eclipse-iceoryx/iceoryx/tree/master/iceoryx_examples/callbacks)

Reviewer: Same for all the other links, we need to link to a fixed version.

```suggestion
[Listener C++ example](https://github.com/eclipse-iceoryx/iceoryx/tree/v2.0.0/iceoryx_examples/callbacks)
```

In a last step we have to release all acquired resources

<!--[geoffrey][iceoryx_examples/callbacks_in_c/ice_c_callbacks_subscriber.c][cleanup]-->
```c
#if !defined(_WIN32)
pthread_join(heartbeatTriggerThread, NULL);

> #endif

Reviewer 2: I wonder whether this 3 lines could be removed without reducing the information for the user.

Author: I asked myself the same question :smiley:

            rightCache.value = *userPayload;
            rightCache.isSet = true;
        }
        printf("received: %d\n", userPayload->counter);
        fflush(stdout);
    }
    // ...
}
```

If both caches are set, we can calculate the sum of both chunks and print them to
the console. To start fresh in the next cycle, we reset the `leftCache` and
the `rightCache` afterward.

<!--[geoffrey][iceoryx_examples/callbacks_in_c/ice_c_callbacks_subscriber.c][[subscriber callback][process data]]-->
```c
void onSampleReceivedCallback(iox_sub_t subscriber)
{
    // ...
    if (leftCache.isSet && rightCache.isSet)
    {
        printf("Received samples from FrontLeft and FrontRight. Sum of %d + %d = %d\n",
               leftCache.value.counter,
               rightCache.value.counter,
               leftCache.value.counter + rightCache.value.counter);
        fflush(stdout);
        leftCache.isSet = false;
        rightCache.isSet = false;
    }
}
```

### Additional context data for callbacks (ice_c_callbacks_with_context_data.c)

> Sometimes you would like to modify data structures within the callback which

Reviewer 3: ```suggestion
Sometimes we would like to modify data structures which
```

> are not globally available. To facilitate this we implemented functions called

Reviewer 3: ```suggestion
are not globally available within the callback. To facilitate this we provide the functions
```

> `iox_listener_attach_***_event_with_context_data` which allow you to provide an

Reviewer 3: ```suggestion
`iox_listener_attach_***_event_with_context_data` which allows to provide an
```

additional void pointer for the callback as second argument.

The following example is a simplified version of the
[ice_c_callbacks_subscriber.c](#ice_c_callbacks_subscriberc) example where we
removed the cyclic heartbeat trigger. The key difference is that we have
a local variable called `counterService` in which we store the `leftCache`
and `rightCache` and we let the callback update that variable directly.

<!--[geoffrey][iceoryx_examples/callbacks_in_c/ice_c_callbacks_with_context_data.c][local variable for caches]-->
```c
CounterService counterService;
counterService.leftCache.isSet = false;
counterService.rightCache.isSet = false;
```

> The callback gets an additional void pointer argument which we cast then to

Reviewer 3: ```suggestion
The callback takes an additional void pointer argument which we cast then to
```
